File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int arr[] = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }
        int dp[] = new int[n];

        int ans = 0;
        dp[0] = 1;
        for (int i = 0; i < n; i++) {
            int cnt = 0;
            for (int j = 0; j < i; j++) {
                if (arr[j] < arr[i]){
                    if (cnt < dp[j])
                        cnt = dp[j];
                }
            }
            dp[i] = cnt + 1;
            ans = Math.max(ans, dp[i]);
        }
        System.out.println(ans);
    }
}

File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.py
import sys
si = sys.stdin.readline

n = int(si().strip())

a = list(map(int, si().split()))
dp = [1 for _ in range(n)]

dp[0] = 1
for i in range(1, n):
	cnt = 0
	for j in range(i):
		if a[j] < a[i]:
			if cnt < dp[j]:
				cnt = dp[j]
	dp[i] = cnt + 1
print(max(dp))